File: trifecta-ios/Trifecta/Core/Models/ServerConfig.swift
import Foundation

struct ServerRuntimeConfig: Decodable, Sendable {
    let providers: [ServerProvider]
}

struct ServerProviderSlashCommand: Decodable, Hashable, Sendable {
    let name: String
    let description: String?
    let input: Input?

    struct Input: Decodable, Hashable, Sendable {
        let hint: String?
    }
}

struct ServerProviderSkill: Decodable, Hashable, Sendable {
    let name: String
    let description: String?
    let shortDescription: String?
}

struct ServerProvider: Decodable, Hashable, Identifiable, Sendable {
    let instanceId: ProviderInstanceID
    let driver: String
    let displayName: String?
    let enabled: Bool
    let installed: Bool
    let status: String
    let auth: ServerProviderAuth
    let models: [ServerProviderModel]
    let showInteractionModeToggle: Bool?
    let slashCommands: [ServerProviderSlashCommand]
    let skills: [ServerProviderSkill]

    var id: ProviderInstanceID { instanceId }

    var isUsable: Bool {
        enabled && installed && auth.status != "unauthenticated"
    }

    var label: String {
        displayName?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
            ? displayName!
            : driver.providerDisplayName
    }

    private enum CodingKeys: String, CodingKey {
        case instanceId, driver, displayName, enabled, installed, status, auth, models
        case showInteractionModeToggle, slashCommands, skills
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        instanceId = try c.decode(ProviderInstanceID.self, forKey: .instanceId)
        driver = try c.decode(String.self, forKey: .driver)
        displayName = try c.decodeIfPresent(String.self, forKey: .displayName)
        enabled = try c.decode(Bool.self, forKey: .enabled)
        installed = try c.decode(Bool.self, forKey: .installed)
        status = try c.decode(String.self, forKey: .status)
        auth = try c.decode(ServerProviderAuth.self, forKey: .auth)
        models = try c.decode([ServerProviderModel].self, forKey: .models)
        showInteractionModeToggle = try c.decodeIfPresent(Bool.self, forKey: .showInteractionModeToggle)
        slashCommands = try c.decodeIfPresent([ServerProviderSlashCommand].self, forKey: .slashCommands) ?? []
        skills = try c.decodeIfPresent([ServerProviderSkill].self, forKey: .skills) ?? []
    }
}

struct ServerProviderAuth: Decodable, Hashable, Sendable {
    let status: String
    let label: String?
    let email: String?
}

struct ServerProviderModel: Decodable, Hashable, Identifiable, Sendable {
    let slug: String
    let name: String
    let shortName: String?
    /// OpenCode: upstream catalog name (Zen, Go, GitHub Copilot, …) — matches desktop `ServerProviderModel.subProvider`.
    let subProvider: String?
    let isCustom: Bool
    /// When the desktop server sends tier/catalog hints (Zen vs Go, etc.).
    let tier: String?
    let catalog: String?
    let bundle: String?
    /// When `false`, the model is hidden from pickers (server marks unusable for this account).
    let eligible: Bool?

    var id: String { slug }
    var label: String { shortName ?? name }

    enum CodingKeys: String, CodingKey {
        case slug, name, shortName, subProvider, isCustom
        case tier, catalog, bundle
        case subscription, routing, channel, offer
        case eligible, available, enabled
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        slug = try c.decode(String.self, forKey: .slug)
        name = try c.decode(String.self, forKey: .name)
        shortName = try c.decodeIfPresent(String.self, forKey: .shortName)
        if let raw = try c.decodeIfPresent(String.self, forKey: .subProvider)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
           !raw.isEmpty {
            subProvider = raw
        } else {
            subProvider = nil
        }
        isCustom = try c.decodeIfPresent(Bool.self, forKey: .isCustom) ?? false

        tier = try Self.decodeFirstString(c, keys: [.tier, .subscription, .routing, .channel, .offer])
        catalog = try c.decodeIfPresent(String.self, forKey: .catalog)
        bundle = try c.decodeIfPresent(String.self, forKey: .bundle)

        eligible = try Self.decodeFirstBool(c, keys: [.eligible, .available, .enabled])
    }

    private static func decodeFirstString(_ c: KeyedDecodingContainer<CodingKeys>, keys: [CodingKeys]) throws -> String? {
        for k in keys {
            if let v = try c.decodeIfPresent(String.self, forKey: k)?
                .trimmingCharacters(in: .whitespacesAndNewlines),
               !v.isEmpty {
                return v
            }
        }
        return nil
    }

    private static func decodeFirstBool(_ c: KeyedDecodingContainer<CodingKeys>, keys: [CodingKeys]) throws -> Bool? {
        for k in keys {
            if let v = try c.decodeIfPresent(Bool.self, forKey: k) { return v }
        }
        return nil
    }
}

/// OpenCode may surface the same logical model through Zen, Go, or direct API routing — used for submenu grouping.
enum OpenCodeRoutingBucket: String, Hashable, Sendable {
    case zen
    case go
    case standard
    case other

    static func fromMetadata(tier: String?, catalog: String?, bundle: String?, subProvider: String? = nil) -> OpenCodeRoutingBucket? {
        let bits = [tier, catalog, bundle, subProvider].compactMap { $0?.lowercased() }
        for raw in bits {
            if raw.contains("zen") { return .zen }
        }
        for raw in bits {
            if raw.contains("google") { continue }
            if raw == "go"
                || raw.hasPrefix("go/")
                || raw.hasSuffix("/go")
                || raw.contains(" opencode go")
                || raw == "opencode-go" {
                return .go
            }
        }
        return nil
    }

    /// Path-based routing, e.g. `zen/openai/gpt-5` vs `go/kimi/...` vs `openai/gpt-5`.
    static func fromSlugPrefix(_ slug: String) -> OpenCodeRoutingBucket? {
        guard let slash = slug.firstIndex(of: "/") else { return nil }
        switch String(slug[..<slash]).lowercased() {
        case "zen": return .zen
        case "go": return .go
        default: return nil
        }
    }

    var sectionSuffix: String {
        switch self {
        case .zen: return "Zen"
        case .go: return "Go"
        case .standard: return "Standard routing"
        case .other: return "Other"
        }
    }

    /// Order for submenu sections in the model picker.
    static let allCasesInOrder: [OpenCodeRoutingBucket] = [.zen, .go, .standard, .other]
}

extension ServerProviderModel {
    func opencodeRoutingBucket() -> OpenCodeRoutingBucket {
        if let fromMeta = OpenCodeRoutingBucket.fromMetadata(tier: tier, catalog: catalog, bundle: bundle, subProvider: subProvider) {
            return fromMeta
        }
        if let fromSlug = OpenCodeRoutingBucket.fromSlugPrefix(slug) {
            return fromSlug
        }
        return .standard
    }
}

extension ServerProvider {
    var defaultModel: String {
        models.first(where: { !$0.isCustom && $0.eligible != false })?.slug
            ?? models.first(where: { $0.eligible != false })?.slug
            ?? models.first?.slug
            ?? defaultModelSlug(for: driver)
    }

    func modelLabel(_ slug: String) -> String {
        models.first { $0.slug == slug }?.label ?? slug
    }

    /// User-visible product name for the integration (Anthropic, Cursor, OpenCode, …).
    var brandDisplayName: String {
        switch driver {
        case "claudeAgent":
            return "Anthropic"
        case "cursor":
            return "Cursor"
        case "opencode":
            return "OpenCode"
        case "openaiChat", "openAIChat", "openai":
            return "OpenAI"
        case "gemini", "googleGemini":
            return "Google"
        default:
            return driver.providerDisplayName
        }
    }

    /// OpenCode uses `upstreamVendor/model` slugs — returns the upstream API vendor when parseable.
    func upstreamVendorLabel(forModelSlug slug: String) -> String? {
        guard driver == "opencode" else { return nil }
        let routed = Self.opencodeUpstreamSlug(slug)
        guard let slash = routed.firstIndex(of: "/") else { return nil }
        let prefix = String(routed[..<slash]).lowercased()
        return Self.opencodeVendorTitle(prefix)
    }

    /// Strips leading `zen/` or `go/` routing prefixes before resolving vendor/model.
    static func opencodeUpstreamSlug(_ slug: String) -> String {
        guard let slash = slug.firstIndex(of: "/") else { return slug }
        switch String(slug[..<slash]).lowercased() {
        case "zen", "go":
            let rest = slug[slug.index(after: slash)...]
            return rest.isEmpty ? slug : String(rest)
        default:
            return slug
        }
    }

    private static func opencodeVendorTitle(_ raw: String) -> String {
        switch raw {
        case "openai": return "OpenAI"
        case "anthropic": return "Anthropic"
        case "google", "gemini": return "Google"
        case "groq": return "Groq"
        case "x-ai", "xai": return "xAI"
        case "mistralai", "mistral": return "Mistral"
        case "deepseek": return "DeepSeek"
        case "meta-llama", "meta": return "Meta"
        case "cohere": return "Cohere"
        default:
            return raw
                .split(separator: "-")
                .map(\.capitalized)
                .joined(separator: " ")
        }
    }

    private func defaultModelSlug(for driver: String) -> String {
        switch driver {
        case "claudeAgent":
            return "claude-sonnet-4-6"
        case "cursor":
            return "auto"
        case "opencode":
            return "openai/gpt-5"
        default:
            return "gpt-5.4"
        }
    }
}

extension ServerRuntimeConfig {
    func modelDisplayLabel(selection: ModelSelection) -> String {
        guard let provider = providers.first(where: { $0.instanceId == selection.instanceId }) else {
            return selection.model
        }
        return provider.modelLabel(selection.model)
    }
}

/// One selectable row in model pickers: a concrete model on a specific provider instance.
struct ModelCatalogEntry: Identifiable, Hashable, Sendable {
    let id: String
    let provider: ServerProvider
    let model: ServerProviderModel

    /// Zen / Go / Standard — for subsection titles and row badges.
    var opencodeBucket: OpenCodeRoutingBucket? {
        guard provider.driver == "opencode" else { return nil }
        return model.opencodeRoutingBucket()
    }

    /// Short badge when metadata names the offer (server may send tier/catalog beyond slug parsing).
    var opencodeOfferBadge: String? {
        guard provider.driver == "opencode" else { return nil }
        if let sp = model.subProvider, !sp.isEmpty { return sp }
        let meta = [model.tier, model.catalog, model.bundle].compactMap { $0?.trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty }
        return meta.first
    }

    static func sortedCatalog(providers: [ServerProvider]) -> [ModelCatalogEntry] {
        let usable = providers.filter(\.isUsable).sorted { $0.label.localizedCaseInsensitiveCompare($1.label) == .orderedAscending }
        var rows: [ModelCatalogEntry] = []
        for p in usable {
            for m in p.models where m.eligible != false {
                rows.append(ModelCatalogEntry(
                    id: "\(p.instanceId.rawValue)|\(m.slug)",
                    provider: p,
                    model: m
                ))
            }
        }
        rows.sort {
            let nameCmp = $0.model.label.localizedCaseInsensitiveCompare($1.model.label)
            if nameCmp != .orderedSame { return nameCmp == .orderedAscending }
            return $0.provider.label.localizedCaseInsensitiveCompare($1.provider.label) == .orderedAscending
        }
        return rows
    }

    /// Secondary line when models are listed in a flat menu (no section headers).
    var pickerSubtitle: String {
        let brand = provider.brandDisplayName
        var parts: [String] = [brand]
        if provider.driver == "opencode", let sp = model.subProvider, !sp.isEmpty {
            parts.append(sp)
        } else if provider.driver == "opencode", let b = opencodeBucket {
            parts.append(b.sectionSuffix)
        }
        if let upstream = provider.upstreamVendorLabel(forModelSlug: model.slug) {
            parts.append(upstream)
        }
        parts.append(provider.label)
        return parts.joined(separator: " · ")
    }
}

/// Models grouped by configured provider instance — OpenCode is further split by Zen / Go / routing bucket.
struct ModelCatalogSection: Identifiable, Sendable {
    /// Stable across subsections (OpenCode uses provider id + bucket).
    let sectionId: String
    let provider: ServerProvider
    let entries: [ModelCatalogEntry]
    /// When set, appended to the header (Zen, Go, Standard routing, …).
    let headerSuffix: String?

    var id: String { sectionId }

    var headerTitle: String {
        let base = "\(provider.brandDisplayName) · \(provider.label)"
        if let headerSuffix {
            return "\(base) · \(headerSuffix)"
        }
        return base
    }

    static func grouped(providers: [ServerProvider]) -> [ModelCatalogSection] {
        let usable = providers
            .filter(\.isUsable)
            .sorted {
                let brandCmp = $0.brandDisplayName.localizedCaseInsensitiveCompare($1.brandDisplayName)
                if brandCmp != .orderedSame { return brandCmp == .orderedAscending }
                return $0.label.localizedCaseInsensitiveCompare($1.label) == .orderedAscending
            }

        var sections: [ModelCatalogSection] = []

        for p in usable {
            let rawEntries: [ModelCatalogEntry] = p.models
                .filter { $0.eligible != false }
                .map { m in
                    ModelCatalogEntry(
                        id: "\(p.instanceId.rawValue)|\(m.slug)",
                        provider: p,
                        model: m
                    )
                }

            guard !rawEntries.isEmpty else { continue }

            if p.driver != "opencode" {
                let sorted = rawEntries.sorted {
                    $0.model.label.localizedCaseInsensitiveCompare($1.model.label) == .orderedAscending
                }
                sections.append(ModelCatalogSection(
                    sectionId: p.instanceId.rawValue,
                    provider: p,
                    entries: sorted,
                    headerSuffix: nil
                ))
                continue
            }

            // Desktop sends `subProvider` per model (OpenCode inventory provider name — Zen, Go, Copilot, …).
            let subProviderKeys = Set(
                rawEntries.compactMap { $0.model.subProvider?.trimmingCharacters(in: .whitespacesAndNewlines) }
                    .filter { !$0.isEmpty }
            )
            if !subProviderKeys.isEmpty {
                let groups = Dictionary(grouping: rawEntries) { entry -> String in
                    let s = entry.model.subProvider?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                    return s.isEmpty ? "__none__" : s
                }
                let orderedKeys = groups.keys.sorted { a, b in
                    if a == "__none__" { return false }
                    if b == "__none__" { return true }
                    return a.localizedCaseInsensitiveCompare(b) == .orderedAscending
                }
                let nonEmptyKeys = orderedKeys.filter { $0 != "__none__" }
                let shouldShowSuffix = nonEmptyKeys.count > 1 || (orderedKeys.contains("__none__") && !nonEmptyKeys.isEmpty)

                for key in orderedKeys {
                    guard var items = groups[key] else { continue }
                    items.sort {
                        $0.model.label.localizedCaseInsensitiveCompare($1.model.label) == .orderedAscending
                    }
                    let label = key == "__none__" ? "Other" : key
                    let suffix: String? = shouldShowSuffix ? label : nil
                    let sid = "\(p.instanceId.rawValue)|\(key)"
                    sections.append(ModelCatalogSection(
                        sectionId: sid,
                        provider: p,
                        entries: items,
                        headerSuffix: suffix
                    ))
                }
            } else {
                let buckets = Dictionary(grouping: rawEntries) { $0.model.opencodeRoutingBucket() }
                let orderedBuckets = OpenCodeRoutingBucket.allCasesInOrder.filter { buckets[$0] != nil }

                let shouldShowSuffix = !(orderedBuckets.count == 1 && orderedBuckets.first == .standard)

                for bucket in orderedBuckets {
                    guard var items = buckets[bucket] else { continue }
                    items.sort {
                        $0.model.label.localizedCaseInsensitiveCompare($1.model.label) == .orderedAscending
                    }
                    let suffix: String? = shouldShowSuffix ? bucket.sectionSuffix : nil

                    let sid = "\(p.instanceId.rawValue)|\(bucket.rawValue)"
                    sections.append(ModelCatalogSection(
                        sectionId: sid,
                        provider: p,
                        entries: items,
                        headerSuffix: suffix
                    ))
                }
            }
        }

        return sections
    }
}

private extension String {
    var providerDisplayName: String {
        replacingOccurrences(of: "([a-z])([A-Z])",
                              with: "$1 $2",
                              options: .regularExpression)
            .replacingOccurrences(of: "[-_]+",
                                  with: " ",
                                  options: .regularExpression)
            .capitalized
    }
}

File: trifecta-android/app/src/main/java/com/belweave/trifecta/core/models/ServerConfig.kt
package com.belweave.trifecta.core.models

import kotlinx.serialization.json.JsonObject

// MARK: - Server runtime config (from /api/v1/runtime-config)

data class ServerProviderAuth(
    val status: String,
    val label: String?,
    val email: String?
) {
    companion object {
        fun fromJson(obj: JsonObject?): ServerProviderAuth =
            ServerProviderAuth(
                status = obj?.str("status") ?: "unauthenticated",
                label = obj?.str("label"),
                email = obj?.str("email")
            )
    }
}

data class ServerProviderModel(
    val slug: String,
    val name: String,
    val shortName: String?,
    val subProvider: String?,
    val isCustom: Boolean,
    val tier: String?,
    val catalog: String?,
    val bundle: String?,
    val eligible: Boolean?
) {
    val id: String get() = slug
    val label: String get() = shortName ?: name

    fun opencodeRoutingBucket(): OpenCodeRoutingBucket {
        OpenCodeRoutingBucket.fromMetadata(tier, catalog, bundle, subProvider)?.let { return it }
        OpenCodeRoutingBucket.fromSlugPrefix(slug)?.let { return it }
        return OpenCodeRoutingBucket.STANDARD
    }

    companion object {
        fun fromJson(obj: JsonObject): ServerProviderModel? {
            val slug = obj.str("slug") ?: return null
            val name = obj.str("name") ?: slug
            val subProviderRaw = obj.str("subProvider")?.trim()
            val tier = firstString(obj, "tier", "subscription", "routing", "channel", "offer")
            val eligible = firstBool(obj, "eligible", "available", "enabled")
            return ServerProviderModel(
                slug = slug,
                name = name,
                shortName = obj.str("shortName"),
                subProvider = subProviderRaw?.takeIf { it.isNotEmpty() },
                isCustom = obj.bool("isCustom") ?: false,
                tier = tier,
                catalog = obj.str("catalog"),
                bundle = obj.str("bundle"),
                eligible = eligible
            )
        }

        private fun firstString(obj: JsonObject, vararg keys: String): String? {
            for (k in keys) {
                obj.str(k)?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }
            }
            return null
        }

        private fun firstBool(obj: JsonObject, vararg keys: String): Boolean? {
            for (k in keys) {
                obj.bool(k)?.let { return it }
            }
            return null
        }
    }
}

data class ServerProviderSlashCommand(
    val name: String,
    val description: String?,
    val inputHint: String?
) {
    companion object {
        fun fromJson(obj: JsonObject): ServerProviderSlashCommand? {
            val name = obj.str("name") ?: return null
            return ServerProviderSlashCommand(
                name = name,
                description = obj.str("description"),
                inputHint = obj.obj("input")?.str("hint")
            )
        }
    }
}

data class ServerProviderSkill(
    val name: String,
    val description: String?,
    val shortDescription: String?
) {
    companion object {
        fun fromJson(obj: JsonObject): ServerProviderSkill? {
            val name = obj.str("name") ?: return null
            return ServerProviderSkill(
                name = name,
                description = obj.str("description"),
                shortDescription = obj.str("shortDescription")
            )
        }
    }
}

data class ServerProvider(
    val instanceId: ProviderInstanceID,
    val driver: String,
    val displayName: String?,
    val enabled: Boolean,
    val installed: Boolean,
    val status: String,
    val auth: ServerProviderAuth,
    val models: List<ServerProviderModel>,
    val showInteractionModeToggle: Boolean?,
    val slashCommands: List<ServerProviderSlashCommand>,
    val skills: List<ServerProviderSkill>
) {
    val id: ProviderInstanceID get() = instanceId

    val isUsable: Boolean
        get() = enabled && installed && auth.status != "unauthenticated"

    val label: String
        get() = displayName?.trim()?.takeIf { it.isNotEmpty() } ?: driver.providerDisplayName()

    val brandDisplayName: String
        get() = when (driver) {
            "claudeAgent" -> "Anthropic"
            "cursor" -> "Cursor"
            "opencode" -> "OpenCode"
            "openaiChat", "openAIChat", "openai" -> "OpenAI"
            "gemini", "googleGemini", "antigravity" -> "Google"
            else -> driver.providerDisplayName()
        }

    val defaultModel: String
        get() = models.firstOrNull { !it.isCustom && it.eligible != false }?.slug
            ?: models.firstOrNull { it.eligible != false }?.slug
            ?: models.firstOrNull()?.slug
            ?: defaultModelSlugFor(driver)

    fun modelLabel(slug: String): String =
        models.firstOrNull { it.slug == slug }?.label ?: slug

    fun upstreamVendorLabel(modelSlug: String): String? {
        if (driver != "opencode") return null
        val routed = opencodeUpstreamSlug(modelSlug)
        val slash = routed.indexOf('/')
        if (slash < 0) return null
        return opencodeVendorTitle(routed.substring(0, slash).lowercase())
    }

    companion object {
        fun fromJson(obj: JsonObject): ServerProvider? {
            val instance = obj.str("instanceId") ?: return null
            val driver = obj.str("driver") ?: return null
            val models = obj.arr("models")?.mapNotNull {
                ServerProviderModel.fromJson(it.asObjectOrNull() ?: return@mapNotNull null)
            } ?: emptyList()
            return ServerProvider(
                instanceId = ProviderInstanceID(instance),
                driver = driver,
                displayName = obj.str("displayName"),
                enabled = obj.bool("enabled") ?: false,
                installed = obj.bool("installed") ?: false,
                status = obj.str("status") ?: "",
                auth = ServerProviderAuth.fromJson(obj.obj("auth")),
                models = models,
                showInteractionModeToggle = obj.bool("showInteractionModeToggle"),
                slashCommands = obj.arr("slashCommands")?.mapNotNull {
                    ServerProviderSlashCommand.fromJson(it.asObjectOrNull() ?: return@mapNotNull null)
                } ?: emptyList(),
                skills = obj.arr("skills")?.mapNotNull {
                    ServerProviderSkill.fromJson(it.asObjectOrNull() ?: return@mapNotNull null)
                } ?: emptyList()
            )
        }

        private fun defaultModelSlugFor(driver: String): String = when (driver) {
            "claudeAgent" -> "claude-sonnet-4-6"
            "cursor" -> "auto"
            "opencode" -> "openai/gpt-5"
            "antigravity" -> "auto"
            else -> "gpt-5.4"
        }

        fun opencodeUpstreamSlug(slug: String): String {
            val slash = slug.indexOf('/')
            if (slash < 0) return slug
            val prefix = slug.substring(0, slash).lowercase()
            return when (prefix) {
                "zen", "go" -> slug.substring(slash + 1).ifEmpty { slug }
                else -> slug
            }
        }

        fun opencodeVendorTitle(raw: String): String = when (raw) {
            "openai" -> "OpenAI"
            "anthropic" -> "Anthropic"
            "google", "gemini" -> "Google"
            "groq" -> "Groq"
            "x-ai", "xai" -> "xAI"
            "mistralai", "mistral" -> "Mistral"
            "deepseek" -> "DeepSeek"
            "meta-llama", "meta" -> "Meta"
            "cohere" -> "Cohere"
            else -> raw.split('-').joinToString(" ") { it.replaceFirstChar { c -> c.uppercase() } }
        }
    }
}

private val camelCaseRegex = Regex("([a-z])([A-Z])")
private val sepRegex = Regex("[-_]+")
private fun String.providerDisplayName(): String =
    this.replace(camelCaseRegex, "$1 $2")
        .replace(sepRegex, " ")
        .split(' ')
        .joinToString(" ") { it.replaceFirstChar { c -> c.uppercase() } }

data class ServerRuntimeConfig(
    val providers: List<ServerProvider>,
    val sshEnabled: Boolean = true,
    val serverPlatformOs: String? = null
) {
    fun provider(instanceId: ProviderInstanceID): ServerProvider? =
        providers.firstOrNull { it.instanceId == instanceId }

    fun modelDisplayLabel(selection: ModelSelection): String {
        val provider = provider(selection.instanceId)
        return provider?.modelLabel(selection.model) ?: selection.model
    }

    companion object {
        fun fromJson(obj: JsonObject): ServerRuntimeConfig {
            val providers = obj.arr("providers")
                ?.mapNotNull { ServerProvider.fromJson(it.asObjectOrNull() ?: return@mapNotNull null) }
                ?: emptyList()
            val environment = obj.obj("environment")
            val sshEnabled = environment?.obj("capabilities")?.bool("ssh") ?: true
            val serverPlatformOs = environment?.obj("platform")?.str("os")
            return ServerRuntimeConfig(
                providers = providers,
                sshEnabled = sshEnabled,
                serverPlatformOs = serverPlatformOs
            )
        }
    }
}

// MARK: - OpenCode routing bucket + catalog grouping

enum class OpenCodeRoutingBucket(val raw: String, val sectionSuffix: String) {
    ZEN("zen", "Zen"),
    GO("go", "Go"),
    STANDARD("standard", "Standard routing"),
    OTHER("other", "Other");

    companion object {
        val allInOrder: List<OpenCodeRoutingBucket> = listOf(ZEN, GO, STANDARD, OTHER)

        fun fromMetadata(
            tier: String?,
            catalog: String?,
            bundle: String?,
            subProvider: String?
        ): OpenCodeRoutingBucket? {
            val bits = listOfNotNull(tier, catalog, bundle, subProvider).map { it.lowercase() }
            for (raw in bits) if (raw.contains("zen")) return ZEN
            for (raw in bits) {
                if (raw.contains("google")) continue
                if (raw == "go" || raw.startsWith("go/") || raw.endsWith("/go")
                    || raw.contains(" opencode go") || raw == "opencode-go"
                ) return GO
            }
            return null
        }

        fun fromSlugPrefix(slug: String): OpenCodeRoutingBucket? {
            val slash = slug.indexOf('/')
            if (slash < 0) return null
            return when (slug.substring(0, slash).lowercase()) {
                "zen" -> ZEN
                "go" -> GO
                else -> null
            }
        }
    }
}

data class ModelCatalogEntry(
    val id: String,
    val provider: ServerProvider,
    val model: ServerProviderModel
) {
    val opencodeBucket: OpenCodeRoutingBucket?
        get() = if (provider.driver == "opencode") model.opencodeRoutingBucket() else null
}

data class ModelCatalogSection(
    val sectionId: String,
    val provider: ServerProvider,
    val entries: List<ModelCatalogEntry>,
    val headerSuffix: String?
) {
    val id: String get() = sectionId

    val headerTitle: String
        get() {
            val base = "${provider.brandDisplayName} · ${provider.label}"
            return if (headerSuffix != null) "$base · $headerSuffix" else base
        }

    companion object {
        fun grouped(providers: List<ServerProvider>): List<ModelCatalogSection> {
            val usable = providers.filter { it.isUsable }
                .sortedWith(
                    Comparator { a, b ->
                        val brandCmp = String.CASE_INSENSITIVE_ORDER.compare(a.brandDisplayName, b.brandDisplayName)
                        if (brandCmp != 0) brandCmp
                        else String.CASE_INSENSITIVE_ORDER.compare(a.label, b.label)
                    }
                )

            val sections = mutableListOf<ModelCatalogSection>()
            for (p in usable) {
                val rawEntries = p.models
                    .filter { it.eligible != false }
                    .map {
                        ModelCatalogEntry(
                            id = "${p.instanceId.rawValue}|${it.slug}",
                            provider = p,
                            model = it
                        )
                    }
                if (rawEntries.isEmpty()) continue

                if (p.driver != "opencode") {
                    val sorted = rawEntries.sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.model.label })
                    sections.add(
                        ModelCatalogSection(
                            sectionId = p.instanceId.rawValue,
                            provider = p,
                            entries = sorted,
                            headerSuffix = null
                        )
                    )
                    continue
                }

                val subProviderKeys = rawEntries
                    .mapNotNull { it.model.subProvider?.trim() }
                    .filter { it.isNotEmpty() }
                    .toSet()

                if (subProviderKeys.isNotEmpty()) {
                    val groups = rawEntries.groupBy {
                        it.model.subProvider?.trim().orEmpty().ifEmpty { "__none__" }
                    }
                    val orderedKeys = groups.keys.sortedWith(Comparator { a, b ->
                        when {
                            a == "__none__" -> 1
                            b == "__none__" -> -1
                            else -> String.CASE_INSENSITIVE_ORDER.compare(a, b)
                        }
                    })
                    val nonEmptyKeys = orderedKeys.filter { it != "__none__" }
                    val showSuffix = nonEmptyKeys.size > 1
                        || (orderedKeys.contains("__none__") && nonEmptyKeys.isNotEmpty())
                    for (key in orderedKeys) {
                        val items = groups[key]?.sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.model.label }) ?: continue
                        val label = if (key == "__none__") "Other" else key
                        sections.add(
                            ModelCatalogSection(
                                sectionId = "${p.instanceId.rawValue}|$key",
                                provider = p,
                                entries = items,
                                headerSuffix = if (showSuffix) label else null
                            )
                        )
                    }
                } else {
                    val buckets = rawEntries.groupBy { it.model.opencodeRoutingBucket() }
                    val orderedBuckets = OpenCodeRoutingBucket.allInOrder.filter { buckets[it] != null }
                    val showSuffix = !(orderedBuckets.size == 1 && orderedBuckets.firstOrNull() == OpenCodeRoutingBucket.STANDARD)
                    for (bucket in orderedBuckets) {
                        val items = buckets[bucket]?.sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.model.label }) ?: continue
                        sections.add(
                            ModelCatalogSection(
                                sectionId = "${p.instanceId.rawValue}|${bucket.raw}",
                                provider = p,
                                entries = items,
                                headerSuffix = if (showSuffix) bucket.sectionSuffix else null
                            )
                        )
                    }
                }
            }
            return sections
        }
    }
}

// MARK: - Pairing environment descriptor

data class EnvironmentDescriptor(
    val policy: String,
    val bootstrapMethods: List<String>,
    val sessionMethods: List<String>,
    val sessionCookieName: String?
) {
    companion object {
        fun decodeLenient(text: String): EnvironmentDescriptor? {
            val obj = runCatching {
                T3Json.parseToJsonElement(text).asObjectOrNull()
            }.getOrNull() ?: return null
            return EnvironmentDescriptor(
                policy = obj.str("policy") ?: "remote-reachable",
                bootstrapMethods = obj.arr("bootstrapMethods")?.mapNotNull { it.stringOrNull() } ?: emptyList(),
                sessionMethods = obj.arr("sessionMethods")?.mapNotNull { it.stringOrNull() } ?: emptyList(),
                sessionCookieName = obj.str("sessionCookieName")
            )
        }
    }
}
